### DISCUSSION TITLE
Access to /dev/dri/renderD128 fails in immich-server docker container

### DISCUSSION THREAD

**Discussion by scherddel** [immich-app/immich#13563] [Q&A] [answered]
**Summary of the Issue:**

I'm experiencing a permission issue when trying to access `/dev/dri/renderD128` inside a Docker container that runs within an unprivileged LXC container on a Proxmox 8.2 host. The host hardware includes a CPU with an integrated GPU (amd, Ryzen 5 7600), which I want to use for transcoding tasks inside the Docker container using VAAPI.

**Environment:**

- **Host:** Proxmox VE 8.2 with an iGPU (amd).
- **LXC Container:**
  - Unprivileged container.
  - Mapped the `render` group GID to align with the host. (see below)
  - Configured to pass through `/dev/dri` devices.
- **Docker Container:**
  - Runs inside the LXC container.
  - Hosts a immich-server, started from the default docker-compose.yml

**Problem Description:**

- The transcoding application inside the Docker container fails with an error when attempting to access `/dev/dri/renderD128`.
```bash
[Nest] 7  - 10/17/2024, 9:15:53 AM   ERROR [Microservices:MediaRepository] ffmpeg version 7.0.2-Jellyfin Copyright (c) 2000-2024 the FFmpeg developers
  built with gcc 12 (Debian 12.2.0-14)
  configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-static --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto=auto --enable-gpl --enable-version3 --enable-shared --enable-gmp --enable-gnutls --enable-chromaprint --enable-opencl --enable-libdrm --enable-libxml2 --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libharfbuzz --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libopenmpt --enable-libdav1d --enable-libsvtav1 --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --arch=amd64 --enable-libshaderc --enable-libplacebo --enable-vulkan --enable-vaapi --enable-amf --enable-libvpl --enable-ffnvcodec --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-nvdec --enable-nvenc
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
Routing option strict to both codec and muxer layer
[AVHWDeviceContext @ 0x52b0c120280] Failed to open /dev/dri/renderD128 as DRM device node.
[AVHWDeviceContext @ 0x52b0c120280] No VA display found for device /dev/dri/renderD128.
Device creation failed: -22.
Failed to set value 'vaapi=accel:/dev/dri/renderD128' for option 'init_hw_device': Invalid argument
Error parsing global options: Invalid argument

[Nest] 7  - 10/17/2024, 9:15:53 AM   ERROR [Microservices:MediaService] Error occurred during transcoding: ffmpeg exited with code 234: Device creation failed: -22.
Failed to set value 'vaapi=accel:/dev/dri/renderD128' for option 'init_hw_device': Invalid argument
Error parsing global options: Invalid argument

[Nest] 7  - 10/17/2024, 9:15:53 AM   ERROR [Microservices:MediaService] Retrying with VAAPI acceleration disabled
```
- Running `cat /dev/dri/renderD128` inside the Docker container also results in "Permission denied".
- If I change the permissions of `/dev/dri/renderD128` on the **host** to `666` (`chmod 666 /dev/dri/renderD128`), the transcoding application works as expected.
- However, changing permissions to `666` is not secure and not an acceptable long-term solution.

**Findings:**

- The device node `/dev/dri/renderD128` inside the Docker container is being overmounted with `tmpfs` and `devtmpfs`, which replaces the host's device node and its permissions.
- Overmounting occurs even without `udev` or `systemd` running inside the container.
- Changing the permissions to `666` on the host allows access because it grants universal read/write permissions, but this is insecure.

- **Device Node Permissions Inside Docker Container:**

  ```bash
  ls -l /dev/dri/renderD128
  ```

  Output:

  ```
  crw-rw---- 1 nobody render 226, 128 Oct 14 21:00 /dev/dri/renderD128
  ```

**Additional Notes:**

- **Security Considerations:**

  - Changing device permissions to `666` on the host is not acceptable due to security risks.
  - Need a solution that maintains proper security while allowing the Docker container to access the iGPU.

**Any help or suggestions would be greatly appreciated.**

### The OS that Immich Server is running on

Promox 8.2.7 running an Alpine 2.19 LXC running docker

### Version of Immich Server

latest

### Version of Immich Mobile App

latest

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
latest
```

### Your .env content

```Shell
latest
```

### Reproduction steps

1. Upload a .MOV file to Immich
2. Transcoding Starts using VAAPI
3. Error occurs and transcoding handled by cpu instead.

**Comment by mertalev** [comment]
Have you tried `group_add`?
```yaml
group_add:
  - render
```
Setting `security_opt` also helps in some cases:
```yaml
security_opt:
  - "label=disable"

**Comment by gmcinalli** [comment]
Hi @scherddel, what CPU you have?

**Comment by scherddel** [comment]
@mertalev Sorry for the late reply, somehow I didnt get notified by your post. I will try your suggestion as well, chances are high that it might work as well since my current solution is to add the immich_server container into gid of the render group via ```user: 0:104``` (104 = render) in the docker-compose.yml immich-server section.

@gmcinalli My server has a ryzen 5 7600. I mentioned it in my initial post, not sure if you didnt see it ;-)

**Reply by scherddel** [reply]
Tried it and it tells me:
```
immich:~/immich-app$ docker compose up -d
[+] Running 5/5
 ✔ Network immich_default             Created                                                                                                                        0.1s 
 ✔ Container immich_postgres          Started                                                                                                                        0.1s 
 ✔ Container immich_machine_learning  Started                                                                                                                        0.1s 
 ✔ Container immich_redis             Started                                                                                                                        0.1s 
 ✔ Container immich_server            Created                                                                                                                        0.0s 
Error response from daemon: Unable to find group render: no matching entries in group file
```
Its worth mentioning that adding a render group manually in the container and then adding root to it also did not resolve anything.

**Reply by scherddel** [reply]
Also 
```
security_opt:
  - "label=disable"
```
did not work.

Back to this solution for now
```
services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    extends:
      file: hwaccel.transcoding.yml
      service: vaapi # set to one of [nvenc, quicksync, rkmpp, vaapi, vaapi-wsl] for accelerated transcoding
    volumes:
      # Do not edit the next line. If you want to change the media storage location on your system, edit the value of UPLOAD_LOCATION in the .env file
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime/Berlin:/etc/localtime:ro
    env_file:
      - .env
    user: 0:104
    ports:
      - '2283:2283'
    depends_on:
      - redis
      - database
    restart: always
    healthcheck:
      disable: false

```

**Reply by mertalev** [reply, marked as answer]
> Tried it and it tells me:
> 
> ```
> immich:~/immich-app$ docker compose up -d
> [+] Running 5/5
>  ✔ Network immich_default             Created                                                                                                                        0.1s 
>  ✔ Container immich_postgres          Started                                                                                                                        0.1s 
>  ✔ Container immich_machine_learning  Started                                                                                                                        0.1s 
>  ✔ Container immich_redis             Started                                                                                                                        0.1s 
>  ✔ Container immich_server            Created                                                                                                                        0.0s 
> Error response from daemon: Unable to find group render: no matching entries in group file
> ```
> 
> Its worth mentioning that adding a render group manually in the container and then adding root to it also did not resolve anything.

Try using the GID for `render` instead of `render`.

**Reply by scherddel** [reply]
Adding the GID worked! I think its a similar approach to adding the 'user 0:104' solution I had but I am glad that this brings the real solution to that issue a little bit closer ;-)

**Reply by elbiozard** [reply]
Hi, I have the same problem. With permission 666 it works, with 660 it doesn't.
What do you mean by "Adding the GID worked! I think its a similar approach to adding the 'user 0:104' solution I had but I am glad that this brings the real solution to that issue a little bit closer ;-)" ?
I tried with these lines and it doesn't work:
devices:
      - /dev/dri:/dev/dri
    user: "0:104"
    group_add:
      - 104
Could you share those lines from docker-compose that worked for you! Thanks

**Reply by scherddel** [reply]
It depends a little bit on your setup but when you exec into your container immich-server and execute the command ls -la /dev/dri what does the output say?

If you like I can guide you through the solution in Discord (I have contacted you there)

**Reply by elbiozard** [reply]
Hi scherddel! Thanks for the help!
I'll pass on what you asked for. I'll check Discord there.

In the HOST:
root@pve:~# ls -la /dev/dri
total 0
drwxr-xr-x  3 root root        100 Oct 22 23:50 .
drwxr-xr-x 18 root root       4740 Oct 22 23:50 ..
drwxr-xr-x  2 root root         80 Oct 22 23:50 by-path
crw-rw----  1 root video  226,   0 Oct 22 23:50 card0
crw-rw-rw-  1 root render 226, 128 Oct 22 23:50 renderD128

On the LCX with Docker:
root@docker:/# ls -la /dev/dri
total 0
drwxr-xr-x 2 root   root          60 Oct 23 00:00 .
drwxr-xr-x 8 root   root         520 Oct 23 01:51 ..
crw-rw-rw- 1 nobody nogroup 226, 128 Oct 22 23:50 renderD128

**Reply by mertalev** [reply]
It would try to use card0 here, so you'd set the GID for video.

**Reply by elbiozard** [reply]
> It would try to use card0 here, so you'd set the GID for video.

Hi @mertalev ! How would this be that you propose? I'm a bit new to this.

**Reply by mertalev** [reply]
So you can run `ls -ln /dev/dri` and see the number where `video` was before. Then use that number when you add the `group_add` section to `immich-server` instead of 104.

**Reply by elbiozard** [reply]
> So you can run `ls -ln /dev/dri` and see the number where `video` was before. Then use that number when you add the `group_add` section to `immich-server` instead of 104.

From Host::
ls -ln /dev/dri
total 0
drwxr-xr-x 2 0   0       80 Oct 22 23:50 by-path
crw-rw---- 1 0  44 226,   0 Oct 22 23:50 card0
crw-rw-rw- 1 0 104 226, 128 Oct 22 23:50 renderD128

From LCX:
ls -ln /dev/dri
total 0
crw-rw-rw- 1 65534 65534 226, 128 Oct 22 23:50 renderD128

I tried with video number (44) adding it in group_add and the same error appears when transcoding.
devices:
    - /dev/dri:/dev/dri
user: "0:104"
group_add:
    - 44 

ERROR:
[Nest] 7  - 10/23/2024, 5:43:07 PM   ERROR [Microservices:MediaRepository] ffmpeg version 7.0.2-Jellyfin Copyright (c) 2000-2024 the FFmpeg developers

  built with gcc 12 (Debian 12.2.0-14)

  configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-static --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto=auto --enable-gpl --enable-version3 --enable-shared --enable-gmp --enable-gnutls --enable-chromaprint --enable-opencl --enable-libdrm --enable-libxml2 --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libharfbuzz --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libopenmpt --enable-libdav1d --enable-libsvtav1 --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --arch=amd64 --enable-libshaderc --enable-libplacebo --enable-vulkan --enable-vaapi --enable-amf --enable-libvpl --enable-ffnvcodec --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-nvdec --enable-nvenc

  libavutil      59.  8.100 / 59.  8.100

  libavcodec     61.  3.100 / 61.  3.100

  libavformat    61.  1.100 / 61.  1.100

  libavdevice    61.  1.100 / 61.  1.100

  libavfilter    10.  1.100 / 10.  1.100

  libswscale      8.  1.100 /  8.  1.100

  libswresample   5.  1.100 /  5.  1.100

  libpostproc    58.  1.100 / 58.  1.100

Routing option strict to both codec and muxer layer

[AVHWDeviceContext @ 0x517b0120280] Failed to open /dev/dri/renderD128 as DRM device node.

[AVHWDeviceContext @ 0x517b0120280] No VA display found for device /dev/dri/renderD128.

Device creation failed: -22.

Failed to set value 'vaapi=accel:/dev/dri/renderD128' for option 'init_hw_device': Invalid argument

Error parsing global options: Invalid argument

[Nest] 7  - 10/23/2024, 5:43:07 PM   ERROR [Microservices:MediaService] Error occurred during transcoding: ffmpeg exited with code 234: Device creation failed: -22.

Failed to set value 'vaapi=accel:/dev/dri/renderD128' for option 'init_hw_device': Invalid argument

Error parsing global options: Invalid argument

[Nest] 7  - 10/23/2024, 5:43:07 PM   ERROR [Microservices:MediaService] Retrying with VAAPI acceleration disabled

[Nest] 7  - 10/23/2024, 5:43:09 PM     LOG [Microservices:MediaService] Successfully encoded 3de4cd69-3d41-479e-9626-289b3fd436f6

**Reply by scherddel** [reply]
Did you map the GID from the proxmox host to the LXC? 

Again I am waiting for you in Discord to guide you though if you like.

**Reply by mertalev** [reply]
It should be 226 instead of 44. You can also remove the `user:` line.

**Reply by elbiozard** [reply]
> Did you map the GID from the proxmox host to the LXC?
> 
> Again I am waiting for you in Discord to guide you though if you like.
@scherddel   I wrote to you a while ago at the account scherddel_16625. My current discord account is biozard22, not the one with the github email.


> It should be 226 instead of 44. You can also remove the `user:` line.
@mertalev  I tried the two options you mentioned and it still fails:
Test 1
devices:
    - /dev/dri:/dev/dri
user: "0:104"
group_add:
    - 226
Test 2
devices:
    - /dev/dri:/dev/dri
#user: "0:104"
group_add:
    - 226

**Reply by gmcinalli** [reply]
@scherddel @elbiozard Have you tried mapping the resource directly in the LXC configuration on Proxmox? You know how to do it, yes? I've been doing this with Plex for years, on a non-privileged LXC, never a problem.
